internal/checks/error.go

```diff
@@ -0,0 +1,219 @@
+}
+
+func NewErrorStats(routingConfig *config.RoutingConfig) ErrorCircuitBreaker {
+	return &ErrorStats{
+		circuitBreaker: NewCircuitBreaker(
+			getErrorsRate(routingConfig),
+			getDetectionWindow(routingConfig),
+			getBanWindow(routingConfig),
+		),
+	}
+}
+
+func NewErrorChecker(
+	upstreamConfig *config.UpstreamConfig,
+	routingConfig *config.RoutingConfig,
+	metricsContainer *metrics.Container,
+	logger *zap.Logger,
+) types.ErrorLatencyChecker {
+	return &ErrorCheck{
+		upstreamConfig:      upstreamConfig,
+		routingConfig:       routingConfig,
+		metricsContainer:    metricsContainer,
+		logger:              logger,
+		errorCircuitBreaker: NewErrorStats(routingConfig),
+	}
+}
+
+func (c *ErrorCheck) isError(httpCode, jsonRPCCode, errorMsg string) bool {
+	if isMatchForPatterns(httpCode, c.routingConfig.Errors.HTTPCodes) ||
+		isMatchForPatterns(jsonRPCCode, c.routingConfig.Errors.JSONRPCCodes) ||
+		isErrorMatches(errorMsg, c.routingConfig.Errors.ErrorStrings) {
+		return true
+	}
+
+	return false
+}
+
+func isMatchForPatterns(responseCode string, patterns []string) bool {
+	if responseCode == "" {
+		return false
+	}
+
+	if len(patterns) == 0 {
+		return true
+	}
+
+	for _, pattern := range patterns {
+		if isMatch(responseCode, pattern) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Returns true iff the response code matches the pattern using ResponseCodeWildcard as the wildcard character.
+func isMatch(responseCode, pattern string) bool {
+	if len(responseCode) != len(pattern) {
+		return false
+	}
+
+	for i, x := range responseCode {
+		y := string(pattern[i])
+
+		if strings.EqualFold(y, string(ResponseCodeWildcard)) {
+			continue
+		}
+
+		if string(x) != y {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isErrorMatches(errorMsg string, errors []string) bool {
+	if errorMsg == "" {
+		return false
+	}
+
+	if len(errors) == 0 {
+		return true
+	}
+
+	for _, errorSubstr := range errors {
+		// TODO(polsar): Add support for regular expression matching.
+		if strings.Contains(errorMsg, errorSubstr) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getErrorsRate(routingConfig *config.RoutingConfig) float64 {
+	if routingConfig != nil && routingConfig.Errors != nil {
+		return routingConfig.Errors.Rate
+	}
+
+	return config.DefaultErrorRate
+}
+
+func (e *ErrorStats) RecordResponse(isError bool) {
+	if isError {
+		e.circuitBreaker.RecordFailure()
+	} else {
+		e.circuitBreaker.RecordSuccess()
+	}
+}
+
+func (e *ErrorStats) IsOpen() bool {
+	// TODO(polsar): We should be able to check `e.circuitBreaker.IsOpen()`,
+	//  but it appears to remain open forever, regardless of the configured delay.
+	//  We also must reset the circuit breaker manually if it is not supposed to be open.
+	isOpen := e.circuitBreaker.RemainingDelay() > 0
+	if !isOpen {
+		e.circuitBreaker.Close()
+	}
+
+	return isOpen
+}
+
+func (c *ErrorCheck) IsPassing([]string) bool {
+	if !c.routingConfig.IsEnhancedRoutingControlDefined() {
```

Comment:
Potential bug: If latency-based routing is ON but error-based routing if OFF, we'll still try to record requests / execute `IsPassing`.  This is because `IsEnhancedRoutingControlDefined` returns true if error OR latency is defined. 

Random idea: should we set a struct-local field like `isCheckEnabled` and check that instead of `IsEnhancedRoutingControlDefined`?